src/plugins/vis_augmenter/public/vega/helpers.ts

```diff
@@ -0,0 +1,324 @@
+ */
+export const addPointInTimeEventsLayersToTable = (
+  datatable: OpenSearchDashboardsDatatable,
+  dimensions: any,
+  visLayers: PointInTimeEventsVisLayer[]
+): OpenSearchDashboardsDatatable => {
+  const augmentedTable = addMissingRowsToTableBounds(datatable, dimensions);
+  const xAxisId = getXAxisId(dimensions, augmentedTable.columns);
+
+  if (!isEmpty(visLayers)) {
+    visLayers.every((visLayer: PointInTimeEventsVisLayer) => {
+      const visLayerColumnId = `${visLayer.pluginResource.id}`;
+      const visLayerColumnName = `${visLayer.pluginResource.name}`;
+      augmentedTable.columns.push({
+        id: visLayerColumnId,
+        name: visLayerColumnName,
+        meta: {
+          type: VIS_LAYER_COLUMN_TYPE,
+        },
+      });
+
+      if (augmentedTable.rows.length === 0) {
+        return false;
+      }
+
+      // if only one row / one datapoint, put all events into this bucket
+      if (augmentedTable.rows.length === 1) {
+        augmentedTable.rows[0] = {
+          ...augmentedTable.rows[0],
+          [visLayerColumnId]: visLayer.events.length,
+        };
+        return false;
+      }
+
+      // Bin the timestamps to the closest x-axis key, adding
+      // an entry for this vis layer ID. Sorting the timestamps first
+      // so that we will only search a particular row value once.
+      // There could be some optimizations, such as binary search + dynamically
+      // changing the bounds, but performance benefits would be very minimal
+      // if any, given the upper bounds limit on n already due to chart constraints.
+      let rowIndex = 0;
+      const sortedTimestamps = visLayer.events
+        .map((event: PointInTimeEvent) => event.timestamp)
+        .sort((n1: number, n2: number) => n1 - n2) as number[];
+
+      if (sortedTimestamps.length > 0) {
+        sortedTimestamps.forEach((timestamp) => {
+          while (rowIndex < augmentedTable.rows.length - 1) {
+            const smallerVal = augmentedTable.rows[rowIndex][xAxisId] as number;
+            const higherVal = augmentedTable.rows[rowIndex + 1][xAxisId] as number;
+            let rowIndexToInsert: number;
+
+            // timestamp is on the left bounds of the chart
+            if (timestamp <= smallerVal) {
+              rowIndexToInsert = rowIndex;
+
+              // timestamp is in between the right 2 buckets. determine which one it is closer to
+            } else if (timestamp <= higherVal) {
+              const smallerValDiff = Math.abs(timestamp - smallerVal);
+              const higherValDiff = Math.abs(timestamp - higherVal);
+              rowIndexToInsert = smallerValDiff <= higherValDiff ? rowIndex : rowIndex + 1;
+            }
+
+            // timestamp is on the right bounds of the chart
+            else if (rowIndex + 1 === augmentedTable.rows.length - 1) {
+              rowIndexToInsert = rowIndex + 1;
+              // timestamp is still too small; traverse to next bucket
```

Comment:
Do we want to show events that exceed the date range of the graph?

Reply:
Right, now that we have `addMissingRowsToTableBounds()` it is true that the bounds of the table will always be equal to (or potentially greater) than the bounds of the chart (see comment about date histogram above addMissingRowsToTableBounds()`).

I'll fix the logic in 2 ways:
1. removing inserting events w/ timestamp < smallest val
2. removing inserting events w/ timestamp > (greatest val + interval)

Reply:
Note that these cases technically shouldn't happen at all, since generated events from plugins should be using the same time context / time range that's provided to the chart, such that results outside of the range should never be found to begin with. But it's good to ensure that in this fn anyways.

Reply:
Added tests for cases of timestamps out of bounds, or on the edge of bounds.